File: bit_manipulation/mix_numbers.py
def mix_numbers(n, m, i, j):
    """
    You are given two numbers, n and m, and two bit positions, i and j.  Write
    a method to insert m into n such that m starts at bit j and ends at bit i.
    You can assume that the bits j through i have enough space to fit all of m.

    >>> mix_numbers(1024, 19, 2, 6) # 10000000000 10011 -> 10001001100
    1100

    >>> mix_numbers(500, 33, 3, 8) # 111110100 100001 -> 100001100
    268
    """
    ones = ~0
    left_mask = ones << (j + 1)
    right_mask = (1 << i) - 1
    mask = left_mask | right_mask
    cleared = n & mask
    shifted = m << i
    return cleared | shifted


if __name__ == '__main__':
    import doctest
    doctest.testmod()

File: src/bit_manipulation/mix_numbers.clj
(ns bit-manipulation.mix-numbers
  (:require [clojure.test :refer :all]))

(defn mix-numbers
  "You are given two numbers, x and y, and two bit positions, i and j.
  Write a method to insert y into x such that y starts at bit j and
  ends at bit n. You can assume that the bits j through n have enough
  space to fit all of y."
  [x y n j]
  (let [right-ones (dec (bit-shift-left 1 n))
        shifted    (bit-shift-left y n)]
    (-> (bit-not 0)
        (bit-shift-left (inc j))
        (bit-or right-ones)
        (bit-and x)
        (bit-or shifted))))

(deftest tests
  (are [x y n j r] (= r (mix-numbers x y n j))
    2r000001      2r11     2 4  2r001101
    2r10000000000 2r10011  2 6  2r10001001100
    2r111110100   2r100001 3 8  2r100001100))

(run-tests)
